src/fetcher.rs

```diff
@@ -1,92 +1,213 @@
-
-    let mut headers = HeaderMap::new();
-    headers.insert("User-Agent", GOSUB_USERAGENT_STRING.parse().unwrap());
-    headers.insert(
-        "Accept",
-        "text/html,application/xhtml+xml,application/xml;q=0.9,*/*;q=0.8".parse().unwrap(),
-    );
-    headers.insert("Accept-Language", "en-US,en;q=0.5".parse().unwrap());
-    // headers.insert("Connection", "keep-alive".parse().unwrap());
-    headers.insert("DNT", "1".parse().unwrap());
-
-    let jar = match SqliteStorage::new("./gosub_cookies.db") {
-        Ok(store) => {
-            info!("successfully created SqliteStorage");
-            Some(CookieJar::new(Arc::new(Mutex::new(store))))
-        }
-        Err(e) => {
-            info!("failed to create SqliteStorage: {:?}", e);
-            None
-        }
-    };
+#[cfg(feature = "proto-http")]
+pub use crate::fetcher::http::{request::HttpRequest, response::HttpResponse, CompleteHttpFetcher, HttpBody, HttpMethod};
 
-    info!("fetching url {}", url);
-    let mut builder = Client::builder()
-        .user_agent(GOSUB_USERAGENT_STRING)
-        .timeout(Duration::from_secs(5))
-        .use_rustls_tls() // For HTTP2
-        .connect_timeout(Duration::from_secs(5))
-        .connection_verbose(true)
-        .read_timeout(Duration::from_secs(5))
-        .brotli(true)
-        .gzip(true)
-        .deflate(true);
-
-    match jar {
-        Some(jar) => {
-            builder = builder.cookie_provider(Arc::new(jar));
-        }
-        None => {
-            info!("no cookie jar");
-        }
+#[cfg(feature = "proto-ftp")]
+pub use crate::fetcher::ftp::{fetcher::FtpFetcher, fetcher::FtpRequest, fetcher::FtpResponse};
+
+#[cfg(feature = "proto-gopher")]
+pub use crate::fetcher::gopher::{fetcher::GopherFetcher, fetcher::GopherRequest, fetcher::GopherResponse};
+
+enum Response {
+    #[cfg(feature = "proto-http")]
+    Http(HttpResponse),
+    #[cfg(feature = "proto-ftp")]
+    Ftp(FtpResponse),
+    #[cfg(feature = "proto-gopher")]
+    Gopher(GopherResponse),
+    // #[cfg(feature = "proto-file")]
+    // File(FileResponse),
+    // #[cfg(feature = "proto-irc")]
+    // Irc(IrcResponse),
+}
+
+pub struct Fetcher {
+    #[cfg(feature = "proto-http")]
+    http_fetcher: CompleteHttpFetcher, // This is the fetcher with the compiled request agent (ie: HttpAgent<ReqwestAgent>)
+    #[cfg(feature = "proto-ftp")]
+    ftp_fetcher: FtpFetcher,
+    #[cfg(feature = "proto-gopher")]
+    gopher_fetcher: GopherFetcher,
+}
+
+impl Fetcher {
+    pub fn protocols_implemented() -> Vec<String> {
+        vec![
+            #[cfg(feature = "proto-http")]
+            "http".to_string(),
+            #[cfg(feature = "proto-ftp")]
+            "ftp".to_string(),
+            // #[cfg(feature = "proto-file")]
+            // "file".to_string(),
+            // #[cfg(feature = "proto-irc")]
+            // "irc".to_string(),
+            #[cfg(feature = "proto-gopher")]
+            "gopher".to_string(),
+        ]
     }
```

Comment:
```suggestion
    pub const fn protocols_implemented() -> &'static [&'static str] {
        &[
            #[cfg(feature = "proto-http")]
            "http",
            #[cfg(feature = "proto-ftp")]
            "ftp",
            // #[cfg(feature = "proto-file")]
            // "file",
            // #[cfg(feature = "proto-irc")]
            // "irc",
            #[cfg(feature = "proto-gopher")]
            "gopher",
        ]
    }
```

Maybe you can even omit the two `'static` lifetimes because the compiler should be able to infer it